### PR TITLE
Feat/pvp  perform duel

### PR DESCRIPTION
**Pull request by RafaelVidaurre**
* Implements and tests the performDuel functionality
* Lints/refactors a bunch of code

### CODE REVIEW

File: test/pvp-arena.js [hunk 1]
```diff
@@ -659,123 +672,493 @@ contract("PvpArena", (accounts) => {
...
   });
 
-  describe("#reRollOpponent", () => {
-    it("fails if character is not dueling", async () => {
-      const character1ID = await createCharacterInPvpTier(accounts[1], 2);
-      await createCharacterInPvpTier(accounts[2], 2);
+  // NOTE: Some tests in this block depend on a deterministic random output.
+  // at this point we can do this in a very limited way, so tests in here are very
+  // fragile, depending both in block time and other tests
+  describe.only("#performDuel", async () => {
```
Comment (Jcalcaldev) [inline test/pvp-arena.js:684]: .

**Review by Jcalcaldev — CHANGES_REQUESTED** [review]

File: contracts/PvpArena.sol [hunk 2]
```diff
@@ -381,6 +473,71 @@ contract PvpArena is Initializable, AccessControlUpgradeable {
         _lastActivityByCharacter[characterID] = block.timestamp;
     }
 
+    function _getCharacterPowerRoll(uint256 characterID, uint8 opponentTrait)
```
Comment (Dvidaurre) [inline contracts/PvpArena.sol:476]: As for now, we are not considering shields in this roll right?

File: contracts/PvpArena.sol [hunk 3]
```diff
@@ -214,8 +234,64 @@ contract PvpArena is Initializable, AccessControlUpgradeable {
...
+            memory bountyDistribution = _getDuelBountyDistribution(attackerID);
+
+        _rewardsByPlayer[winner] = _rewardsByPlayer[winner].add(
+            bountyDistribution.winnerReward
+        );
+        _fightersByCharacter[loserID].wager = _fightersByCharacter[loserID]
+            .wager
+            .sub(bountyDistribution.loserPayment);
+
+        if (_fightersByCharacter[loserID].wager == 0) {
```
[the remark this answers is not in the document]
Comment (RafaelVidaurre) [reply contracts/PvpArena.sol:282]: No, that's on maneki.

Wager can never be < 0, these are unsigned integers (unsigned = "without a sign")

File: contracts/PvpArena.sol [hunk 4]
```diff
@@ -381,6 +473,71 @@ contract PvpArena is Initializable, AccessControlUpgradeable {
...
+    {
+        // TODO:
+        // - [ ] consider shield
+        uint8 trait = characters.getTrait(characterID);
+        uint24 basePower = characters.getPower(characterID);
+        uint256 weaponID = _fightersByCharacter[characterID].weaponID;
+        uint256 seed = randoms.getRandomSeed(msg.sender);
+
+        (
+            ,
```
[the remark this answers is not in the document]
Comment (RafaelVidaurre) [reply contracts/PvpArena.sol:489]: this is response destructuring. Functions in solidity can return tuples of an arbitrary amount of values. I don't need the first one here so a comma signals to skip it (in solidity declaring vars in memory uses a "slot" per function, reaching 16 of these causes an error)

File: test/pvp-arena.js (hunk 1 above)
Comment (RafaelVidaurre) [reply test/pvp-arena.js:684]: noice catch

File: contracts/PvpArena.sol (hunk 2 above)
Comment (RafaelVidaurre) [reply contracts/PvpArena.sol:476]: yup, there was a comment somewhere

**Pull request merged**